File: src/test/kotlin/com/boclips/event/service/testsupport/VideoFactory.kt
package com.boclips.event.service.testsupport

import com.boclips.event.service.testsupport.SubjectFactory.createSubjects
import com.boclips.eventbus.domain.AgeRange
import com.boclips.eventbus.domain.video.ContentPartner
import com.boclips.eventbus.domain.video.Dimensions
import com.boclips.eventbus.domain.video.PlaybackProviderType
import com.boclips.eventbus.domain.video.Video
import com.boclips.eventbus.domain.video.VideoAsset
import com.boclips.eventbus.domain.video.VideoId
import com.boclips.eventbus.domain.video.VideoType
import java.time.ZonedDateTime

object VideoFactory {

    fun createVideo(
        id: String = "",
        title: String = "",
        contentPartnerName: String = "",
        playbackProviderType: PlaybackProviderType = PlaybackProviderType.KALTURA,
        subjectNames: List<String> = emptyList(),
        ageRange: AgeRange = AgeRange(),
        durationSeconds: Int = 180,
        ingestedAt: ZonedDateTime = ZonedDateTime.now(),
        type: VideoType = VideoType.INSTRUCTIONAL,
        originalDimensions: Dimensions? = Dimensions(640, 480),
        assets: List<VideoAsset>? = listOf()
    ): Video {
        return Video
            .builder()
            .id(VideoId(id))
            .ingestedAt(ingestedAt)
            .title(title)
            .contentPartner(ContentPartner.of(contentPartnerName))
            .playbackProviderType(playbackProviderType)
            .subjects(createSubjects(subjectNames))
            .ageRange(ageRange)
            .durationSeconds(durationSeconds)
            .type(type)
            .originalDimensions(originalDimensions)
            .assets(assets)
            .build()
    }
}

File: event-aggregator/src/test/scala/com/boclips/event/aggregator/testsupport/testfactories/VideoFactory.scala
package com.boclips.event.aggregator.testsupport.testfactories

import java.time.{Duration, LocalDate, ZonedDateTime}

import com.boclips.event.aggregator.domain.model.contentpartners.ChannelId
import com.boclips.event.aggregator.domain.model.videos._
import com.boclips.event.aggregator.domain.model.{videos, _}

object VideoFactory {

  def createVideo(
                   id: String = "123",
                   releasedOn: LocalDate = LocalDate.now(),
                   ingestedAt: ZonedDateTime = ZonedDateTime.now(),
                   title: String = "title",
                   channelId: String = "channel id",
                   contentType: Option[String] = None,
                   subjects: List[String] = List("English"),
                   playbackProvider: String = "KALTURA",
                   playbackId: String = "playbackId",
                   assets: List[VideoAsset] = Nil,
                   originalDimensions: Option[Dimensions] = None,
                   ageRange: AgeRange = AgeRange(Some(5), Some(7)),
                   duration: Duration = Duration.ofSeconds(180),
                   promoted: Boolean = false,
                   topics: List[VideoTopic] = Nil,
                   keywords: List[String] = Nil,
                   sourceVideoReference: Option[String] = None,
                 ): Video = {
    videos.Video(
      id = VideoId(id),
      releasedOn = releasedOn,
      ingestedAt = ingestedAt,
      title = title,
      contentType = contentType,
      channelId = ChannelId(channelId),
      playbackProvider = playbackProvider,
      playbackId = playbackId,
      assets = assets,
      originalDimensions = originalDimensions,
      subjects = subjects.map(name => Subject(name)),
      ageRange = ageRange,
      duration = duration,
      promoted = promoted,
      topics = topics,
      keywords = keywords,
      sourceVideoReference = sourceVideoReference,
    )
  }

  def createVideoAsset(
                        sizeKb: Int = 1000,
                        dimensions: Dimensions = Dimensions(1920, 1080),
                        bitrateKbps: Int = 100
                      ): VideoAsset = {
    VideoAsset(
      sizeKb = sizeKb,
      dimensions = dimensions,
      bitrateKbps = bitrateKbps
    )
  }

  def createStorageCharge(
                           videoId: VideoId = VideoId("video-id"),
                           periodStart: LocalDate = LocalDate.now(),
                           periodEnd: LocalDate = LocalDate.now(),
                           valueGbp: Double = 1
                         ): VideoStorageCharge = {
    VideoStorageCharge(
      videoId = videoId,
      periodStart = periodStart,
      periodEnd = periodEnd,
      valueGbp = valueGbp,
    )
  }
}
